type-metadata/src/type_id.rs

```diff
@@ -0,0 +1,146 @@
+use derive_more::From;
+use serde::Serialize;
+
+pub trait HasTypeId {
+	fn type_id() -> TypeId;
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug, Serialize)]
+pub struct Namespace {
+	segments: Vec<&'static str>,
+}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Debug, Serialize)]
+pub struct InvalidNamespace;
```

Comment:
We could make `InvalidNamespace` as enum, to provide different kind of reasons about why invalid.